sstar/stats.py

```diff
@@ -0,0 +1,277 @@
+# Apache License Version 2.0
+# Copyright 2022 Xin Huang
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import numpy as np
+import scipy.stats as sps
+from scipy.spatial import distance_matrix
+
+
+
+
+def cal_n_ton(tgt_gt):
+    """
+    Description:
+        Calculates individual frequency spetra for haplotypes.
+    Arguments:
+        tgt_gt numpy.ndarray: Genotype matrix from the target population.
+    Returns:
+        spectra numpy.ndarray: Individual frequency spectra for haplotypes.
+    """
+
+
+    try:
+        mut_num, hap_num = tgt_gt.shape
+        iv = np.ones((hap_num, 1))
+        counts = tgt_gt*np.matmul(tgt_gt, iv)
+        spectra = np.array([np.bincount(counts[:,idx].astype('int8'), minlength=hap_num+1) for idx in range(hap_num)])
+        # ArchIE does not count non-segragating sites
+        spectra[:,0] = 0
+    except Exception as e: 
+        print(e)
+        import sys
+        print("shape of input genotype matrix:")
+        print(tgt_gt.shape)
+        print("unique values")
+        print(np.unique(tgt_gt))
+        np.set_printoptions(threshold=sys.maxsize)
+        print("input tgt_gt")
+        print(tgt_gt)
+
+
+    return spectra
+
+
+def cal_n_ton_v2a(tgt_gt):
+    """
+    Description:
+        Calculates individual frequency spetra for haplotypes - entries greater than 1 changed to 1.
+    Arguments:
+        tgt_gt numpy.ndarray: Genotype matrix from the target population.
+    Returns:
+        spectra numpy.ndarray: Individual frequency spectra for haplotypes.
+    """
+
+    mut_num, hap_num = tgt_gt.shape
+    iv = np.ones((hap_num, 1))
+    counts = tgt_gt*np.matmul(tgt_gt, iv)
+    spectra = np.array([np.bincount(np.array(counts[:,idx] > 0).astype('int8'), minlength=hap_num+1) for idx in range(hap_num)])
```

Comment:
This line is not correct. The spectra here only contain "single-ton", because `counts[:,idx] > 0` returns an array only containing 0s and 1s.

Reply:
Therefore, what this line founds is the total number of mutations that a haplotype contains.

Reply:
The problem was caused by `astype('int8')`, because the default type is `float64`, after converting to `int8`, negative numbers may occur. To fix it, we can change `astype('int8')` to `astype('int64')`

Reply:
The negative number problem was caused by astype('int8'), because the range of int8 is from -128 to 127, when simulating large sample sizes (e.g. 100 diploid target individuals), negative numbers may occur. To fix it, we can change astype('int8') to astype('int64')

Reply:
The negative number problem was caused by `astype('int8')`, because the range of `int8` is from -128 to 127, when simulating large sample sizes (e.g. 100 diploid target individuals), negative numbers may occur. To fix it, we can change `astype('int8')` to `astype('int64')`